packages/filesystem/src/browser/file-tree/file-tree-widget.ts

```diff
@@ -56,4 +63,105 @@ export class FileTreeWidget extends TreeWidget {
+        super.onAfterAttach(msg);
+        this.addEventListener(this.node, 'dragenter', event => this.handleDragEnterEvent(event));
+        this.addEventListener(this.node, 'dragover', event => this.handleDragOverEvent(event));
+        this.addEventListener(this.node, 'dragleave', event => this.handleLeaveEvent(event));
+        this.addEventListener(this.node, 'drop', event => this.handleDropEvent(this.model.root, event, false));
+    }
+
+    protected createNodeAttributes(node: ITreeNode, props: NodeProps): ElementAttrs {
+        const elementAttrs = super.createNodeAttributes(node, props);
+        if (DirNode.is(node)) {
+            // Folder, so it could be a dropzone
+            return {
+                ...elementAttrs,
+                ondrop: event => this.handleDropEvent(node, event, true),
+            };
+        }
+        return elementAttrs;
+    }
+
+    protected handleDragEnterEvent(event: DragEvent): void {
+        event.preventDefault();
+        event.stopPropagation();
+        const target = event.target as HTMLElement;
+        if (target.parentElement) {
+            target.parentElement.classList.add(activeDropZone);
+        }
+    }
+
+    protected handleDragOverEvent(event: DragEvent): void {
+        event.preventDefault();
+    }
+
+    protected handleLeaveEvent(event: DragEvent): void {
+        event.preventDefault();
+        event.stopPropagation();
+        const target = event.target as HTMLElement;
+        if (target.parentElement) {
+            target.parentElement.classList.remove(activeDropZone);
+        }
+    }
+
+    protected handleDropEvent(node: ITreeNode | undefined, event: DragEvent, dropAllowed: boolean): void {
+        event.stopPropagation();
+        event.preventDefault();
+        if (dropAllowed) {
+            event.dataTransfer.dropEffect = 'copy'; // Explicitly show this is a copy.
+            const target = event.target as HTMLElement;
+            if (target.parentElement) {
+                target.parentElement.classList.remove(activeDropZone);
+            }
+            this.dropHandler(event, node);
+        }
+    }
+
+    protected dropHandler(ev: DragEvent, node: ITreeNode | undefined) {
+        ev.preventDefault();
+        const data = ev.dataTransfer.items;
+
+        for (let i = 0; i < data.length; i += 1) {
+            if (data[i].kind === 'file') {
+                const entry: WebKitEntry = data[i].webkitGetAsEntry();
+                if (node !== undefined) {
+                    if (entry.isDirectory) {
+                        this.handleDirectory(entry, node.id);
+                    } else {
+                        const f: File | null = data[i].getAsFile();
+                        if (f !== null) {
+                            this.model.upload(node.id, f, f.name);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    protected handleDirectory(webDirEntry: WebKitEntry, dest: string) {
```

Comment:
Is `WebKitEntry` part of the standardized browser API? I do not know; I am just asking. I found [this](https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement/webkitEntries). Thanks!